#45933 Injects a base plugin with default UI method implementation

josh-t opened the pull request:
This commit corresponds to the changes in https://github.com/shotgunsoftware/tk-core/pull/568. 

A new publish plugin base class hook is added to the `hooks` folder for discoverability. But it is not configured. Instead, the publisher itself extracts the class and supplies it to the `create_hook_instance` method in core as the plugins are instantiated. This means we can define the default task settings UI in a hook without requiring studios to modify their configuration. Any plugins that do not implement the UI methods will pick this up automatically. 

There is more work to do here to style, but I wanted to get this in front of you guys in conjunction with the core changes to see the whole thing.

robblau commented on python/tk_multi_publish2/processing/plugin.py:
```diff
@@ -129,6 +139,49 @@ def __init__(self, name, path, settings, logger):
+            # don't include a .py here. it's a legacy core thing to add it
+            # automatically
+            "plugin_base"
+        )
```
Is there a way to pass something like "{self}/hooks/plugin_base.py" to a method and get the class or module back?

josh-t replied:
Yes! This is much cleaner now. Done!

robblau commented on hooks/plugin_base.py:
```diff
@@ -0,0 +1,133 @@
+    whatever way you want.
+    """
+
+    def create_settings_widget(self, parent):
```
Is this now duplicated code from the app?  If so, any way to call it from the app or have the app use the code from here?

josh-t replied:
Good call. I made changes to keep an instance of the base plugin around and use that whenever we want to generate a default UI for a plugin that doesn't implement the methods. This required a minor hack to inject the current plugin's "description" onto the base plugin instance, but seems like a reasonable tradeoff to not have to duplicate the default description UI logic. Here's the relevant bits: 

* https://github.com/shotgunsoftware/tk-multi-publish2/pull/52/files#diff-6da8c710ddfe37588ac2fc105dda81ccR176
* https://github.com/shotgunsoftware/tk-multi-publish2/pull/52/files#diff-6da8c710ddfe37588ac2fc105dda81ccR338

Let me know what you think @robblau.

robblau replied:
I think it is fine.  Given that the app is the one providing the hook base class I think it is perfectly fine for the app to have specific knowledge about what it does.

manneohrstrom commented on hooks/plugin_base.py:
```diff
@@ -0,0 +1,133 @@
+        # the default implementation does not show any editable widgets, so this
+        # is a no-op. this method is required to be defined in order for the
+        # custom UI to show up in the app
+        pass
```
to complete the interface, should we add accept-validate etc methods too? With comments?

josh-t replied:
Yeah, I was thinking about that. This seems like the file to use when generating the docs. I will update.

manneohrstrom commented on python/tk_multi_publish2/dialog.py:
```diff
@@ -178,6 +173,13 @@ def __init__(self, parent=None):
+        # defined there to create the default task UI. This will prevent us
+        # from having to create it over and over as tasks are selected in
+        # the UI.
+        self._plugin_base_hook = self._bundle.create_hook_instance(
```
❤️

manneohrstrom commented on python/tk_multi_publish2/dialog.py:
```diff
@@ -319,35 +321,32 @@ def _update_task_details_ui(self, new_task_selection=None):
+            # a bit of a hack here, but the base hook requires a description in
+            # order to function. we'll simply monky patch it here to be able
+            # to generate the UI.
+            self._plugin_base_hook.description = new_task_selection.plugin.description
```
not sure i totally follow, but could this be done via inheritance? doesn't `new_task_selection.plugin.description` inherit from plugin base hook?

josh-t replied:
This is hopefully true, but for studios that haven't updated core, that base class injection will fail silently, and the regular `Hook` base class will be used. That means the hook may not have the custom UI. So we keep this reusable hook instance that does implement the methods and we use that to generate the widget. It requires a description, so we essentially monkey patch it with the selected tasks' description.

manneohrstrom replied:
but if they are using `HookBaseClass = sgtk.get_hook_baseclass()` they will just get this new interface auto-injected, right? Sorry, i am probably missing some detail here! :)

josh-t replied:
Only if they have the new core. At least the way we're currently doing the injection. Lets discuss.

manneohrstrom commented on python/tk_multi_publish2/processing/plugin.py:
```diff
@@ -129,6 +139,44 @@ def __init__(self, name, path, settings, logger):
+            # retrieve the class for the plugin base hook. this hook lives in
+            # the app itself, so we can refer to it with the hook expression
+            # that would be used if configured
+            plugin_base_class = self._bundle.get_hook_class(plugin_base_file)
```
i think this is over complicated. I don't think we should think of the "base plugin" as a hook but as a normal class - probably not even keep it in the hooks folder - that way we can simply import it and we don't need this (imho a little weird!) method `bundle.get_hook_class`. The base plugin is a fixed interface and a class that is part of the publish code base and not a hook. Does that make sense?

josh-t replied:
I originally wanted to do it this way, but my thought was that keeping it in the hook folder helped keep it visible to clients and made it discoverable.

manneohrstrom replied:
we should also think about sphinx docs - and hook docs in particular - just realised that by having a hook interface like this, and placing the class in the "normal" code, we can super easily sphinx document it!!! 😍

josh-t replied:
Yeah. was thinking the same.

josh-t commented:
@manneohrstrom, @robblau 

I should not have removed the `try/except` I had in before to account for older cores that do not support the `base_class` argument when creating hook instances. I've fixed that and added some new logic to patch plugins created without the base class injection. The code will now bind the base plugin's custom ui methods to the hook instance. This ensures that we always have plugins with the custom ui methods implemented. The patching code is a bit low level, so would be good to get input on it. Here's the specific commit to avoid the many other changes in this PR:  https://github.com/shotgunsoftware/tk-multi-publish2/pull/52/commits/77c497bccdef0157525222fe9c73a5d9b7449d40

Thanks!

manneohrstrom commented:
77c497b looks complicated but the only other way i can imagine it working is if we set the min required core to be 0.18.12x (whatever it is). Given how new the publisher is, and our new awesome release scheme, i think we should consider that approach.

manneohrstrom commented on python/tk_multi_publish2/processing/plugin.py:
```diff
@@ -37,15 +37,23 @@ def __init__(self, path, settings, logger):
+        """
+        Create the plugin's hook instance. Subclasses can reimplement for more
+        sophisticated hook instantiation.
+        :return:
```
nit: looks like it's missing some docstring infos.

josh-t replied:
addressed.
